File: 2016/1taxicab.py
fn = '1.dat'

with open(fn) as f:
    moves = f.read().strip().split(', ')

compass = [[0, 1], [1, 0], [0, -1], [-1, 0]]
x, y, heading = 0, 0, 0

for move in moves:
    heading = (heading + 1 if move[0] == 'R' else heading - 1) % 4
    dx, dy = compass[heading]
    dist = int(move[1:])
    x += dx * dist
    y += dy * dist

print (f"Part 1: {abs(x) + abs(y)}")

move_set = set()

x, y, heading = 0, 0, 0

try:
    for move in moves:
        move_set.add((x, y))
        heading = (heading + 1 if move[0] == 'R' else heading - 1) % 4
        dx, dy = compass[heading]
        dist = int(move[1:])
        for _ in range(dist):
            x += dx
            y += dy
            if (x, y) in move_set:
                raise StopIteration
            move_set.add((x, y))
except:
    pass

print (f"Part 2: {abs(x) + abs(y)}")

File: 2016/1taxicab.clj
(ns aoc.2016.1
  (:require [clojure.string :as str]))

(def compass [[0 1] [1 0] [0 -1] [-1 0]])

(defn new-heading [heading move]
  (mod (if (= (first move) \R) (inc heading) (dec heading)) 4))

(defn travel [heading arr]
  (if (empty? arr) []
      (into (vec (repeat (Integer/parseInt (subs (first arr) 1)) (new-heading heading (first arr))))
            (travel (new-heading heading (first arr)) (rest arr)))))

(defn journey [loc moves]
  (concat [loc]
          (if (empty? moves) []
              (journey [(+ (loc 0) ((first moves) 0)) (+ (loc 1) ((first moves) 1))] (rest moves)))))

(defn look_for_duplicates [path]
  (if (.contains (rest path) (first path)) (first path) (look_for_duplicates (rest path))))

(def head-seq (travel 0 (str/split (read-line) #",\s")))

(def path (journey [0 0] (map vector
                              (for [h head-seq] ((compass h) 0))
                              (for [h head-seq] ((compass h) 1)))))

(println "Part 1:" (+ (Math/abs ((last path) 0)) (Math/abs ((last path) 1))))

(def dupe (look_for_duplicates path))

(println "Part 2:" (+ (Math/abs (dupe 0)) (Math/abs (dupe 1))))
